src/compiler/binder.rs

```diff
@@ -63,19 +71,113 @@ impl BinderType {
         *self.parent.try_write().unwrap() = parent;
     }
 
+    fn container(&self) -> Rc<Node> {
+        self.container.try_read().unwrap().as_ref().unwrap().clone()
+    }
+
+    fn maybe_container(&self) -> Option<Rc<Node>> {
+        self.container
+            .try_read()
+            .unwrap()
+            .as_ref()
+            .map(Clone::clone)
+    }
+
+    fn set_container(&self, container: Option<Rc<Node>>) {
+        *self.container.try_write().unwrap() = container;
+    }
+
+    #[allow(non_snake_case)]
+    fn Symbol(&self) -> fn(SymbolFlags, __String) -> Symbol {
+        self.Symbol.try_read().unwrap().unwrap()
+    }
+
+    #[allow(non_snake_case)]
+    fn set_Symbol(&self, Symbol: fn(SymbolFlags, __String) -> Symbol) {
+        *self.Symbol.try_write().unwrap() = Some(Symbol);
+    }
+
     fn bind_source_file(&self, f: Rc<Node>) {
         self.set_file(Some(f.clone()));
 
+        self.set_Symbol(object_allocator.get_symbol_constructor());
+
         if true {
             self.bind(Some(self.file()));
         }
 
         self.set_file(None);
         self.set_parent(None);
+        self.set_container(None);
+    }
+
+    fn create_symbol(&self, flags: SymbolFlags, name: __String) -> Symbol {
+        self.Symbol()(flags, name)
+    }
+
+    fn add_declaration_to_symbol(&self, symbol: Rc<Symbol>, node: Rc<Node /*Declaration*/>) {
+        node.set_symbol(symbol.clone());
```

Comment:
`Symbol`'s get stored in a symbol table but also as a field on `Node`'s, so I'm wrapping them in `Rc`'s and storing a weak reference in `BaseNode`'s `symbol` field